src/pam_hbac_ldap_compat.c

```diff
@@ -0,0 +1,264 @@
+/*
+    Copyright (C) 2016 Jakub Hrozek <jakub.hrozek@posteo.se>
+
+    This program is free software; you can redistribute it and/or modify
+    it under the terms of the GNU Lesser General Public License as published by
+    the Free Software Foundation; either version 3 of the License, or
+    (at your option) any later version.
+
+    This program is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU Lesser General Public License for more details.
+
+    You should have received a copy of the GNU Lesser General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "pam_hbac_compat.h"
+#include "pam_hbac_ldap.h"
+
+#include <ctype.h>
+
+#if !defined(HAVE_LDAP_STR2DN) || defined(COMPAT_LDAP_UNIT_TESTS)
+static int rdn_keyval(const char *exploded_rdn_value,
+                      char **_key, char **_val)
+{
+    size_t i;
+    char *eq_sign;
+    const char *rdn_start;
+    const char *val_start;
+
+    if (exploded_rdn_value == NULL) {
+        return EINVAL;
+    }
+
+    /* Skip leading whitespace before the key */
+    rdn_start = exploded_rdn_value;
+    for (i = 0; isspace(rdn_start[i]); i++);
+
+    rdn_start += i;
+    if (*rdn_start == '\0') {
+        return EINVAL;
+    }
+
+    eq_sign = strchr(rdn_start, '=');
+    if (eq_sign == NULL) {
+        return EINVAL;
+    }
+
+    val_start = eq_sign+1;
+    /* Skip leading whitespace before the value */
+    for (i =0; isspace(val_start[i]); i++);
+
+    /* Empty value? */
+    val_start += i;
+    if (*val_start == '\0') {
+        return EINVAL;
+    }
+
+    *_key = strndup(rdn_start, eq_sign - rdn_start);
+    *_val = strdup(val_start);
+
+    return 0;
+}
+
+static LDAPAVA *rdn2ava(const char *rdn)
+{
+    char **exploded_rdn = NULL;
+    LDAPAVA *ava;
+    char *key = NULL;
+    char *val = NULL;
+    int ret;
+
+    /* For single-valued RDNs, this is a noop, for multiple-valued, we just
+     * hope for the best. Because we don't support deny rules, the user
+     * might be not allowed access at worst
+     */
+    exploded_rdn = ldap_explode_rdn(rdn, 0);
+    if (exploded_rdn == NULL) {
+        return NULL;
+    }
+
+    ava = malloc(sizeof(LDAPAVA));
+    if (ava == NULL) {
+        ldap_value_free(exploded_rdn);
+        return NULL;
+    }
+
+    /* Since the server is IPA, we can ignore anything but the simplest,
+     * single-valued RDNs, we know IPA doesn't use those
+     */
+    ret = rdn_keyval(exploded_rdn[0], &key, &val);
+    ldap_value_free(exploded_rdn);
+    if (ret != 0 || key == NULL || val == NULL) {
+        free(key);
```

Comment:
On Mon, May 23, 2016 at 07:05:10AM -0700, Pavel Reichl wrote:
> > +        return NULL;
> > +    }
> > +
> > +    ava = malloc(sizeof(LDAPAVA));
> > +    if (ava == NULL) {
> > +        ldap_value_free(exploded_rdn);
> > +        return NULL;
> > +    }
> > +
> > +    /* Since the server is IPA, we can ignore anything but the simplest,
> > +     * single-valued RDNs, we know IPA doesn't use those
> > +     */
> > +    ret = rdn_keyval(exploded_rdn[0], &key, &val);
> > +    ldap_value_free(exploded_rdn);
> > +    if (ret != 0 || key == NULL || val == NULL) {
> > +        free(key);
> 
> Should we free ava here as well?

Fixed.